File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Negocio/BO/crm/ClienteBOImpl.cs
using ProyectoProgramacion3Model.Model;
using ProyectoProgramacion3Model.Model.crm;
using ProyectoProgramacion3Persistencia.DAO.crm;

namespace ProyectoProgramacion3Negocio.BO.crm;

public class ClienteBOImpl : BaseBO, IClienteBO
{
    private readonly IClienteDao clienteDao;

    public ClienteBOImpl()
    {
        clienteDao = new ClienteDaoImpl();
    }

    public List<Cliente> Listar() => clienteDao.LeerTodos();

    public Cliente? Obtener(int id)
    {
        ValidarIdPositivo(id, "id del cliente");
        return clienteDao.Leer(id);
    }

    public void Eliminar(int id)
    {
        ValidarIdPositivo(id, "id del cliente");
        if (!clienteDao.Eliminar(id))
        {
            throw new InvalidOperationException($"No se pudo eliminar el cliente con id: {id}");
        }
    }

    public void Guardar(Cliente modelo, Estado estado)
    {
        ValidarCliente(modelo);
        ValidarEstado(estado);

        if (estado == Estado.Nuevo)
        {
            var id = clienteDao.Crear(modelo);
            if (id <= 0)
            {
                throw new InvalidOperationException("No se pudo registrar el nuevo cliente");
            }

            modelo.idCliente = id;
        }
        else if (estado == Estado.Modificado)
        {
            ValidarIdPositivo(modelo.idCliente, "id del cliente");
            if (!clienteDao.Actualizar(modelo))
            {
                throw new InvalidOperationException($"No se pudo actualizar el cliente con id: {modelo.idCliente}");
            }
        }
    }

    private static void ValidarCliente(Cliente modelo)
    {
        ArgumentNullException.ThrowIfNull(modelo);
        ValidarTextoObligatorio(modelo.nombres, "nombres del cliente");
        ValidarTextoObligatorio(modelo.apellidos, "apellidos del cliente");
        ValidarTextoObligatorio(modelo.numeroDocumento, "numero de documento");

        if (!string.IsNullOrWhiteSpace(modelo.correo) && !modelo.correo.Contains('@'))
        {
            throw new ArgumentException("El formato del correo electronico es invalido");
        }

        ValidarDateOnly(modelo.fechaRegistro, "fecha de registro");
    }
}

File: Backend/ProyectoPro3Negocio/src/main/java/pe/edu/pucp/proyectopro3/bo/crm/ClienteBOImpl.java
package pe.edu.pucp.proyectopro3.bo.crm;

import pe.edu.pucp.proyectopro3.bo.BaseBO;
import pe.edu.pucp.proyectopro3.dao.crm.ClienteDAO;
import pe.edu.pucp.proyectopro3.dao.crm.ClienteDAOImpl;
import pe.edu.pucp.proyectopro3.modelo.Estado;
import pe.edu.pucp.proyectopro3.modelo.crm.Cliente;

import java.util.List;
import java.util.Objects;

public class ClienteBOImpl extends BaseBO implements ClienteBO {
    private final ClienteDAO clienteDao;

    public ClienteBOImpl() {
        // Se inicializa el DAO correspondiente
        this.clienteDao = new ClienteDAOImpl();
    }

    @Override
    public List<Cliente> listar() {
        return this.clienteDao.leerTodos();
    }

    @Override
    public Cliente obtener(int id) {
        validarIdPositivo(id, "id del cliente");
        return this.clienteDao.leer(id);
    }

    @Override
    public void eliminar(int id) {
        validarIdPositivo(id, "id del cliente");
        if (!this.clienteDao.eliminar(id)) {
            throw new IllegalStateException("No se pudo eliminar el cliente con id: " + id);
        }
    }

    @Override
    public void guardar(Cliente modelo, Estado estado) {
        validarCliente(modelo);
        validarEstado(estado);

        if (estado == Estado.Nuevo) {
            int id = this.clienteDao.crear(modelo);
            if (id <= 0) {
                throw new IllegalStateException("No se pudo registrar el nuevo cliente");
            }
            modelo.setIdCliente(id);
        }
        else if (estado == Estado.Modificado) {
            validarIdPositivo(modelo.getIdCliente(), "id del cliente");
            if (!this.clienteDao.actualizar(modelo)) {
                throw new IllegalStateException("No se pudo actualizar el cliente con id: " + modelo.getIdCliente());
            }
        }
    }

    private void validarCliente(Cliente modelo) {
        Objects.requireNonNull(modelo, "El objeto Cliente no puede ser nulo");
        validarTextoObligatorio(modelo.getNombres(), "nombres del cliente");
        validarTextoObligatorio(modelo.getApellidos(), "apellidos del cliente");
        validarTextoObligatorio(modelo.getNumeroDocumento(), "número de documento");

        // Validación adicional de negocio: correo electrónico
        if (modelo.getCorreo() != null && !modelo.getCorreo().contains("@")) {
            throw new IllegalArgumentException("El formato del correo electrónico es inválido");
        }

        // La fecha de registro no puede ser nula
        if (modelo.getFechaRegistro() == null) {
            throw new IllegalArgumentException("La fecha de registro es obligatoria");
        }
    }
}
